### PR TITLE
fixes #11862 - validating that casted hash is valid

### PR DESCRIPTION
**Pull request by unorthodoxgeek**

### CODE REVIEW

**Comment by dLobatog** [comment]
@unorthodoxgeek Looks like you're hitting some issue with Psych on 1.9.3 :/

**Comment by unorthodoxgeek** [comment]
ugh, why doesn't the rescue catch that exception on 1.9.3? the same exception is thrown on 2.x and it gets caught correctly. doh :facepunch:

**Comment by domcleal** [comment]
Catch the specific error, since Psych::SyntaxError inherits from StandardError in 2+, but is a ScriptError in 1.9.3, so won't be caught by a general `rescue`.

**Comment by orrabin** [comment]
@unorthodoxgeek this needs rebase.

File: app/services/foreman/parameters/validator.rb [hunk 1]
```diff
@@ -43,6 +45,17 @@ def validate_list
         true
       end
 
+      def validate_hash
+        return true if value.is_a?(Hash)
```
Comment (domcleal) [inline app/services/foreman/parameters/validator.rb]: Could a YAML or JSON key not have an Array value still?

It's unclear to me if there's any test coverage of this conditional.

File: app/services/foreman/parameters/validator.rb [hunk 2]
```diff
@@ -43,6 +45,17 @@ def validate_list
...
+        return false unless value.is_a?(String)
+        return false if value.blank?
```
Comment (domcleal) [inline app/services/foreman/parameters/validator.rb]: missing test coverage of these conditionals?

**Comment by unorthodoxgeek** [comment]
@domcleal - fixed according to your comments, thanks :+1:

File: app/services/foreman/parameters/validator.rb [hunk 3]
```diff
@@ -13,6 +13,10 @@ def validate!
           validate_regexp
         when "list", "array"
           validate_list
+        when "hash"
```
Comment (domcleal) [inline app/services/foreman/parameters/validator.rb]: Sorry, I need to take a step back.  I can't reproduce the original issue described in http://projects.theforeman.org/issues/11862#note-2, and secondly, I can't see how this code is related.

a) Every time I try and reproduce the original issue, I get a validation message:

<pre>
2015-10-22T15:12:02 [app] [I] Started PUT "/puppetclasses/45-activemq" for 127.0.0.1 at 2015-10-22 15:12:02 +0100
2015-10-22T15:12:02 [app] [I] Processing by PuppetclassesController#update as HTML
2015-10-22T15:12:02 [app] [I]   Parameters: {"utf8"=>"✓", "authenticity_token"=>"WA5GDpEyXQDdN06vepyFOBARRgQtB495MWgVbORPMtE=", "puppetclass"=>{"hostgroup_ids"=>["", "39"], "class_params_attributes"=>{....}, "lookup_keys_attributes"=>{"0"=>{"_destroy"=>"false", "key"=>"test2", "description"=>"foobar", "key_type"=>"hash", "default_value"=>"[FI
LTERED]", "validator_type"=>"", "path"=>"fqdn\r\nhostgroup\r\nos\r\ndomain", "merge_overrides"=>"0", "id"=>"220"}}}, "environment_filter"=>"", "commit"=>"»Submit«", "id"=>"45-activemq"}
...
2015-10-22T15:12:24 [app] [I] Failed to save: »Default value« »is invalid«
</pre>


Note that the key_type is hash, I have no validator and no matchers.  The default value should be cast only, no validation.  The error seems to be recorded fine, no exception, and is shown in the UI correctly next to the default value.

b) This particular line is actually in the validator class.  If you trace where it's called from, it's in LookupKey#validate_default_value and the `@options[:type]` is actually LookupKey#validator_type.  That's only ever going to be none, regexp or list.  hash, yaml, json are key_types, not validator_types.

Comment (unorthodoxgeek) [inline app/services/foreman/parameters/validator.rb]: reproduction:
select "hash" as type of value
input { "foo" => "bar } into default value (or lookup values)
see everything crash and burn

Comment (domcleal) [inline app/services/foreman/parameters/validator.rb]: I still see the "is invalid" error after entering your suggestion as the default value, and get the same logging as I pasted above.

Comment (unorthodoxgeek) [inline app/services/foreman/parameters/validator.rb]: it's actually a regression of sorts, because now for some reason the value gets turned into yaml and breaks even more...

edit: this bug is only in ruby 1.9.3, nothing changed in develop

Comment (unorthodoxgeek) [inline app/services/foreman/parameters/validator.rb]: screen shot 2015-11-09 at 5 26 34 pm
this is how it currently looks (after two invalid submissions)

Comment (unorthodoxgeek) [inline app/services/foreman/parameters/validator.rb]: ok, this bug is only reproducible in ruby 1.9.3, in which case this error message appears:
screen shot 2015-11-10 at 1 02 58 pm

**Comment by unorthodoxgeek** [comment]
@domcleal - this error occurs only in ruby 1.9.3, in ruby 2+ it works as expected. I've taken the liberty of fixing a tiny bug in the behaviour - when the default value is not valid, we shouldn't encode it as yaml, we should just show what was inputted by the user so they can fix it.

in ruby 2.0+ this PR only fixes that tiny validation issue.

File: app/services/foreman/parameters/caster.rb [hunk 4]
```diff
@@ -110,10 +110,14 @@ def cast_yaml
...
       def load_yaml_or_json
+        self.class.load_yaml_or_json(value)
+      end
+
+      def self.load_yaml_or_json(value)
         return value unless value.is_a? String
         begin
           JSON.load value
-        rescue
+        rescue Psych::SyntaxError, JSON::ParserError
```
Comment (domcleal) [inline app/services/foreman/parameters/caster.rb]: Rescuing the Psych error here doesn't make sense, it has to be around the YAML.load, surely?  I still see the problem on this branch, under Ruby 1.9.3 emanating from the YAML.load line.

File: app/services/foreman/parameters/validator.rb (hunk 3 above)
Comment (domcleal) [inline app/services/foreman/parameters/validator.rb]: Right, I see it now, thanks.  That comes back to the additional rescue correctly catching the right exception under 1.9.3.

Now I'm left with this point:

> b) This particular line is actually in the validator class. If you trace where it's called from, it's in LookupKey#validate_default_value and the `@options[:type]` is actually LookupKey#validator_type. That's only ever going to be none, regexp or list. hash, yaml, json are key_types, not validator_types.

I don't see why these changes are required, or that they make any difference.

Comment (domcleal) [inline app/services/foreman/parameters/validator.rb]: Could you please address this point?

File: app/services/foreman/parameters/validator.rb [hunk 5]
```diff
@@ -43,6 +43,23 @@ def validate_list
         true
       end
 
+      def validate_hash
```
Comment (domcleal) [inline app/services/foreman/parameters/validator.rb]: It looks like you've removed code that calls this, but not these methods.  Are they unused now?

**Comment by domcleal** [comment]
I thought this PR had tests originally, have they been removed?  Could you add some?

I'm not really sure what's changing in this PR on each update, so [please do comment](https://groups.google.com/d/msg/foreman-dev/poyjF8Ig5UE/jaFyZVGOsEUJ).

**Comment by unorthodoxgeek** [comment]
@domcleal - you pointed out that what I did was pointless and never got called, so I removed the code and left only the rescue and the small ui fix (when invalid, use the invalid value without casting it)
don't see the point in tests, as this is throwing an error in 1.9.3 without this fix, and nothing substantial changed except the 1.9.3 fix.

**Comment by domcleal** [comment]
> don't see the point in tests, as this is throwing an error in 1.9.3 without this fix, and nothing substantial changed except the 1.9.3 fix.

Tests are important to prevent regressions, we require them for changes wherever possible.  This looks quite easy to test too, particularly if it's throwing an exception.

**Comment by unorthodoxgeek** [comment]
@domcleal - added test which makes sure that the right error is thrown (fails on Ruby 1.9.3 without the code changes, doesn't fail on other rubies)

File: app/services/foreman/parameters/caster.rb [hunk 6]
```diff
@@ -110,11 +110,15 @@ def cast_yaml
       end
 
       def load_yaml_or_json
+        self.class.load_yaml_or_json(value)
+      end
+
+      def self.load_yaml_or_json(value)
```
Comment (domcleal) [inline app/services/foreman/parameters/caster.rb]: This can stay as an instance method?  It doesn't appear to be called as a class method anywhere.

**Comment by domcleal** [comment]
The PR should also contain tests for:
1. The error (still) thrown from the caster is caught in the `validate_and_cast_value` / `cast_default_value` methods in the two models it's used in.  It's changed from a Runtime to a Standard error, but I can't see coverage that this is caught (please point it out if there already).
2. Coverage of the `*_before_type_cast` changes on validation failure.

**Comment by unorthodoxgeek** [comment]
regarding tests:
1. has tests - caster_test.rb:81 lookup_key_test.rb:303 lookup_value_test.rb:107
2. added tests.

File: test/unit/lookup_value_test.rb [hunk 7]
```diff
@@ -111,6 +111,12 @@ def valid_attrs2
     assert_equal lk.value_before_type_cast, "[{\"foo\":\"bar\"},{\"baz\":\"qux\"},\"baz\"]"
   end
 
+  test "should not cast string if object invalid" do
+    lk = LookupValue.new(:value => '{"foo" => "bar"}', :match => "hostgroup=Common", :lookup_key => lookup_keys(:seven))
+    refute lk.save
```
Comment (dLobatog) [inline test/unit/lookup_value_test.rb]: `.valid?`

File: test/unit/lookup_key_test.rb [hunk 8]
```diff
@@ -192,6 +192,15 @@ def test_parameters_multiple_paths
       assert_equal '["<%= @host.name %>"]', param.default_value
       assert_equal default, param.default_value_before_type_cast
     end
+
+    test "when invalid, just returns the invalid value" do
+      val = '{"foo" => "bar"}'
+      param = FactoryGirl.build(:puppetclass_lookup_key, :as_smart_class_param,
+                                :override => true, :key_type => 'hash',
+                                :default_value => val, :puppetclass => puppetclasses(:one))
+      refute param.save
```
Comment (dLobatog) [inline test/unit/lookup_key_test.rb]: `.valid?`

File: app/models/lookup_keys/lookup_key.rb [hunk 9]
```diff
@@ -124,6 +124,7 @@ def reject_invalid_lookup_values(attributes)
   end
 
   def default_value_before_type_cast
+    return read_attribute(:default_value) if errors.present? && errors[:default_value].present?
```
Comment (dLobatog) [inline app/models/lookup_keys/lookup_key.rb]: I think `errors.present?` is redundant here, all ActiveRecord objects will respond true to `.errors.present?` even if there's no error.

File: app/models/lookup_value.rb [hunk 10]
```diff
@@ -45,6 +45,7 @@ def name
   end
 
   def value_before_type_cast
+    return read_attribute(:value) if errors.present? && errors[:value].present?
```
Comment (dLobatog) [inline app/models/lookup_value.rb]: Same thing as on lookup_key, under what circumstances checking `errors.present?` is necessary?

**Comment by dLobatog** [comment]
@unorthodoxgeek Few nitpicks inline, I think the integration test failure was just a coincidence?

File: app/models/lookup_value.rb (hunk 10 above)
Comment (unorthodoxgeek) [inline app/models/lookup_value.rb]: if errors is blank, won't calling errors[:value] throw a NoMethodError :[] for nil?

Comment (unorthodoxgeek) [inline app/models/lookup_value.rb]: ok. errors is ActiveModel::Errors
it does respond to #present? with false when no errors, but I'll remove it :smile:

**Comment by unorthodoxgeek** [comment]
@dLobatog - fixed

File: app/views/lookup_keys/_value.html.erb [hunk 11]
```diff
@@ -9,7 +9,7 @@
     <%= text_field_tag '', nil, :class => 'matcher_value' %>
   </td>
   <td>
-    <%= hidden_value_field(f, :value, f.object.value, f.object.use_puppet_default, :'data-property' => 'value') %>
+    <%= hidden_value_field(f, :value, f.object.value_before_type_cast, f.object.use_puppet_default, :'data-property' => 'value') %>
```
Comment (domcleal) [inline app/views/lookup_keys/_value.html.erb:12]: This doesn't seem to work on the Puppet class edit form, as when entering an invalid value for a hash smart variable's matcher and submitting the form, it reloads with the `--- ! '...'` variant.  It looks like the `value` passed into the helper is only used when creating hidden value fields, not visible fields.

Comment (unorthodoxgeek) [inline app/views/lookup_keys/_value.html.erb:12]: this isn't a regression.
if the value is invalid (but key's default value is valid), it acts like it should.

if key is invalid, the value doesn't even get its validations run (thanks DHH)
this seems like a AR/AM bug, and it's present in develop. we should fix it, but let's do it in a different task.

Comment (domcleal) [inline app/views/lookup_keys/_value.html.erb:12]: Right, so this is just intending to fix the case when loading the form and a hidden value is ticked.

I quite agree about using different tasks - the combining of various smart class variable bugs in this PR has made it difficult to review, as the contents keep changing.  This seems more related to #6007 rather than #11862.

**Comment by domcleal** [comment]
Merged as 706de3c, thanks @unorthodoxgeek.
